### PR TITLE
Added support for additional start copy parameters

### PR DESCRIPTION
**Pull request by gapra-msft**

### CODE REVIEW

File: sdk/storage/azure-storage-file-share/src/test/java/com/azure/storage/file/share/APISpec.groovy [hunk 1]
```diff
@@ -39,7 +39,7 @@ class APISpec extends Specification {
     String connectionString
 
     // If debugging is enabled, recordings cannot run as there can only be one proxy at a time.
-    static boolean enableDebugging = false
+    static boolean enableDebugging = true
```
Comment (rickle-msft) [inline sdk/storage/azure-storage-file-share/src/test/java/com/azure/storage/file/share/APISpec.groovy]: revert

File: sdk/storage/azure-storage-file-share/src/test/java/com/azure/storage/file/share/FileAPITests.groovy [hunk 2]
```diff
@@ -424,6 +425,59 @@ class FileAPITests extends APISpec {
...
+        when:
+        SyncPoller<ShareFileCopyInfo, Void> poller = primaryFileClient.beginCopy(sourceURL, null, null,
+            PermissionCopyModeType.SOURCE, true, false, null, null)
+
+        def pollResponse = poller.poll()
+
+        then:
+        assert pollResponse.getValue().getCopyId() != null
+    }
+
```
Comment (rickle-msft) [inline sdk/storage/azure-storage-file-share/src/test/java/com/azure/storage/file/share/FileAPITests.groovy]: I think these tests can be parameterized and collapsed into one test. You also shouldn't need an assert statement in a then block

Comment (rickle-msft) [reply sdk/storage/azure-storage-file-share/src/test/java/com/azure/storage/file/share/FileAPITests.groovy]: The async tests can also be collapsed

Comment (gapra-msft) [reply sdk/storage/azure-storage-file-share/src/test/java/com/azure/storage/file/share/FileAPITests.groovy]: Collapsed

File: sdk/storage/azure-storage-file-share/src/main/java/com/azure/storage/file/share/ShareFileAsyncClient.java [hunk 3]
```diff
@@ -241,26 +243,83 @@ public ShareServiceVersion getServiceVersion() {
...
+
+        FileSmbProperties finalSmbProperties = smbProperties == null ? new FileSmbProperties() : smbProperties;
+
+        String filePermissionKey = finalSmbProperties.getFilePermissionKey();
+
+        String fileAttributes = finalSmbProperties.setNtfsFileAttributes(FileConstants.FILE_ATTRIBUTES_NONE);
+        String fileCreationTime = FileSmbProperties.parseFileSMBDate(finalSmbProperties.getFileCreationTime());
+        String fileLastWriteTime = FileSmbProperties.parseFileSMBDate(finalSmbProperties.getFileLastWriteTime());
+
+        if (filePermissionCopyMode == PermissionCopyModeType.SOURCE) {
```
Comment (xiafu-msft) [inline sdk/storage/azure-storage-file-share/src/main/java/com/azure/storage/file/share/ShareFileAsyncClient.java]: do we want to discuss when filePermissionCopyMode is null?
if if permission_key or permission_key is set and filePermissionCopyMode is null, probably we want to throw exception?

File: sdk/storage/azure-storage-file-share/src/test/java/com/azure/storage/file/share/FileAPITests.groovy [hunk 4]
```diff
@@ -424,6 +425,37 @@ class FileAPITests extends APISpec {
...
+        when:
+        SyncPoller<ShareFileCopyInfo, Void> poller = primaryFileClient.beginCopy(sourceURL, smbProperties,
+            setFilePermission ? filePermission : null, permissionType, ignoreReadOnly,
+            setArchiveAttribute, null, null)
+
+        def pollResponse = poller.poll()
+
+        then:
+        pollResponse.getValue().getCopyId() != null
```
Comment (xiafu-msft) [inline sdk/storage/azure-storage-file-share/src/test/java/com/azure/storage/file/share/FileAPITests.groovy]: Is it convenient to validate the SMB attributes are the same as the set ones also the acl is copied from source?
It is to validate the headers are exactly sent out from client side.

File: sdk/storage/azure-storage-file-share/src/main/java/com/azure/storage/file/share/ShareFileAsyncClient.java (hunk 3 above)
Comment (gapra-msft) [reply sdk/storage/azure-storage-file-share/src/main/java/com/azure/storage/file/share/ShareFileAsyncClient.java]: In the doc it says it's an optional param

File: sdk/storage/azure-storage-file-share/src/test/java/com/azure/storage/file/share/FileAPITests.groovy (hunk 4 above)
Comment (gapra-msft) [reply sdk/storage/azure-storage-file-share/src/test/java/com/azure/storage/file/share/FileAPITests.groovy]: I can add those

**Pull request merged**